src/main/java/org/rumbledb/compiler/VariableDependenciesVisitor.java

```diff
@@ -0,0 +1,466 @@
+        // TODO;
+        return defaultAction(expression, argument);
+    }
+
+    @Override
+    public Void visitQuantifiedExpression(QuantifiedExpression expression, Void argument) {
+        List<QuantifiedExpressionVar> var = expression.getVariables();
+        addInputVariableDependencies(
+            expression,
+            this.inputVariableDependencies.get(expression.getEvaluationExpression())
+        );
+        for (QuantifiedExpressionVar v : var) {
+            visit(v.getExpression(), null);
+            removeInputVariableDependency(expression, v.getVariableName());
+        }
+        for (QuantifiedExpressionVar v : var) {
+            addInputVariableDependencies(expression, this.inputVariableDependencies.get(v.getExpression()));
+        }
+        return null;
+    }
+
+    @Override
+    public Void visitTypeSwitchExpression(TypeSwitchExpression expression, Void argument) {
+        visit(expression.getTestCondition(), null);
+        for (TypeswitchCase v : expression.getCases()) {
+            visit(v.getReturnExpression(), null);
+            addInputVariableDependencies(expression, this.inputVariableDependencies.get(v.getReturnExpression()));
+            removeInputVariableDependency(expression, v.getVariableName());
+        }
+        addInputVariableDependencies(expression, this.inputVariableDependencies.get(expression.getTestCondition()));
+        return null;
+    }
+
+    @Override
+    public Void visitNamedFunctionRef(NamedFunctionReferenceExpression expression, Void argument) {
+        addInputVariableDependency(expression, "Function " + expression.getIdentifier());
+        return null;
+    }
+
+    @Override
+    public Void visitFunctionCall(FunctionCallExpression expression, Void argument) {
+        addInputVariableDependency(
+            expression,
+            "Function " + expression.getFunctionName() + "#" + expression.getArguments().size()
+        );
+        for (Expression e : expression.getArguments()) {
+            if (e != null) {
+                visit(e, null);
+                addInputVariableDependencies(expression, this.inputVariableDependencies.get(e));
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public Void visitDynamicFunctionCallExpression(DynamicFunctionCallExpression expression, Void argument) {
+        visit(expression.getMainExpression(), null);
+        addInputVariableDependencies(expression, this.inputVariableDependencies.get(expression.getMainExpression()));
+        for (Expression e : expression.getArguments()) {
+            if (e == null) {
+                continue;
+            }
+            visit(e, null);
+            addInputVariableDependencies(expression, this.inputVariableDependencies.get(e));
+        }
+        return null;
+    }
+
+    @Override
+    public Void visitProlog(Prolog prolog, Void argument) {
+        Map<String, Node> nameToNodeMap = new TreeMap<>();
```

Comment:
I would highly recommend extracting the contents of this method to a few functions for readability

Reply:
DONE